app/controllers/FileValidationController.scala

```diff
@@ -16,25 +16,79 @@
+import repositories.SessionRepository
 import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
-import views.html.FileValidationView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
 
 class FileValidationController @Inject() (
   override val messagesApi: MessagesApi,
   identify: IdentifierAction,
   getData: DataRetrievalAction,
-  requireData: DataRequiredAction,
+  val sessionRepository: SessionRepository,
   val controllerComponents: MessagesControllerComponents,
-  view: FileValidationView
-) extends FrontendBaseController
+  upscanConnector: UpscanConnector,
+  requireData: DataRequiredAction,
+  validationConnector: ValidationConnector,
+  navigator: Navigator
+)(implicit ec: ExecutionContext)
+    extends FrontendBaseController
     with I18nSupport {
 
-  def onPageLoad: Action[AnyContent] = (identify andThen getData() andThen requireData) {
+  def onPageLoad(): Action[AnyContent] = (identify andThen getData() andThen requireData).async {
     implicit request =>
-      Ok(view())
+      {
+        for {
+          uploadId       <- getUploadId(request.userAnswers)
+          uploadSessions <- upscanConnector.getUploadDetails(uploadId)
+          (fileName, upScanUrl) = getDownloadUrl(uploadSessions)
+          validation: Option[Either[ValidationErrors, Boolean]] <- validationConnector.sendForValidation(upScanUrl)
+        } yield validation match {
+          case Some(Right(_)) =>
+            for {
+              updatedAnswers        <- Future.fromTry(request.userAnswers.set(ValidXMLPage, fileName))
+              updatedAnswersWithURL <- Future.fromTry(updatedAnswers.set(URLPage, upScanUrl))
+              _                     <- sessionRepository.set(updatedAnswersWithURL)
+            } yield Redirect(navigator.nextPage(ValidXMLPage, NormalMode, updatedAnswers))
+
+          case Some(Left(ValidationErrors(errors, _))) =>
+            for {
+              updatedAnswers           <- Future.fromTry(UserAnswers(request.userId).set(InvalidXMLPage, fileName))
+              updatedAnswersWithErrors <- Future.fromTry(updatedAnswers.set(GenericErrorPage, errors))
+              _                        <- sessionRepository.set(updatedAnswersWithErrors)
+            } yield Redirect(navigator.nextPage(InvalidXMLPage, NormalMode, updatedAnswers))
+
+          case _ =>
+            for {
+              updatedAnswers <- Future.fromTry(UserAnswers(request.userId).set(InvalidXMLPage, fileName))
+              _              <- sessionRepository.set(updatedAnswers)
+            } yield Redirect(navigator.nextPage(InvalidXMLPage, NormalMode, updatedAnswers))
+        }
+      }.flatten
   }
+
+  private def getUploadId(userAnswers: UserAnswers): Future[UploadId] =
+    userAnswers.get(UploadIDPage) match {
+      case Some(uploadId) => Future.successful(uploadId)
+      case None           => throw new RuntimeException("Cannot find uploadId")
+    }
```

Comment:
This  can be 
  Future.fromTry {
      Try {
        userAnswers.get(UploadIDPage).getOrElse(throw new RuntimeException("Cannot find uploadId"))
      }
    }

Reply:
Is there any benefit in doing that? I think it's more readable using pattern matching

Reply:
Your exception is not in the same future so you have to mix block async together since you are throwing an exception, it won't get related to the same thread. it will be associated to the main thread